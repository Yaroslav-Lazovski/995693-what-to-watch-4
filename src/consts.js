export const MAX_SIMILAR_MOVIES = 4;

export const GENRES = {
  ALL: `All genres`,
  COMEDIAN: `Comedian`,
  CRIME: `Crime`,
  DOCUMENTARY: `Documentary`,
  DRAMA: `Drama`,
  HORROR: `Horror`,
  KIDS_FAMILY: `Kids & Family`,
  ROMANCE: `Romance`,
  SCI_FI: `Sci-Fi`,
  THRILLER: `Thriller`,
};

export const MovieRating = {
  BAD: 0,
  NORMAL: 3,
  GOOD: 5,
  VERY_GOOD: 8,
  AWESOME: 10
};

export const RatingType = {
  BAD: `Bad`,
  NORMAL: `Normal`,
  GOOD: `Good`,
  VERY_GOOD: `Very good`,
  AWESOME: `Awesome`,
};

export const TabType = {
  OVERVIEW: `Overview`,
  DETAILS: `Details`,
  REVIEWS: `Reviews`
};

export const AuthorizationStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`,
};

export const MAX_MOVIES_IN_LIST = 8;

export const AppRoute = {
  ADD_REVIEW: `/review`,
  FILM: `/films`,
  LOGIN: `/login`,
  MY_LIST: `/mylist`,
  PLAYER: `/player`,
  ROOT: `/`
};

export const dateOptions = {
  year: `numeric`,
  month: `long`,
  day: `numeric`
};

export const dateInnerOptions = {
  year: `numeric`,
  month: `numeric`,
  day: `numeric`
};

export const Time = {
  SECONDS_IN_MINUTE: 60,
  SECONDS_IN_HOUR: 3600
};
