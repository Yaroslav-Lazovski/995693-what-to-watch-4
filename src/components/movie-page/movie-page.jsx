import React, {PureComponent, Fragment} from "react";
import PropTypes from "prop-types";
import {connect} from "react-redux";
import {Link} from "react-router-dom";

import MoviesList from "../movies-list/movies-list.jsx";
import MovieOverview from "../movie-overview/movie-overview.jsx";
import MovieDetails from "../movie-details/movie-details.jsx";
import MovieReviews from "../movie-reviews/movie-reviews.jsx";
import Header from "../header/header.jsx";
import Footer from "../footer/footer.jsx";
import AddMyList from "../add-my-list/add-my-list.jsx";
import withActiveCard from "../../hocs/with-active-card.js";

import {ActionCreator} from "../../reducer/state/state";
import {getSelectedMovie} from "../../reducer/state/selectors.js";
import {getAuthorizationStatus} from "../../reducer/user/selectors.js";
import {getMovies} from "../../reducer/data/selectors.js";
import {getLoadingFavoriteMovie} from '../../reducer/data/selectors';
import {Operation as DataOperation} from '../../reducer/data/data';
import {AppRoute, TabType, MAX_SIMILAR_MOVIES} from "../../consts.js";

const MoviesListWrapped = withActiveCard(MoviesList);


export class MoviePage extends PureComponent {
  constructor(props) {
    super(props);

    this._renderActiveTab = this._renderActiveTab.bind(this);
  }

  componentDidMount() {
    const {movie, setActiveMovie} = this.props;

    setActiveMovie(movie);
  }

  componentDidUpdate() {
    const {movie, setActiveMovie} = this.props;

    setActiveMovie(movie);
  }

  _renderActiveTab() {
    const {movie: {genre, year, ratingScore, ratingCount, description, director, starring, runTime}, activeTab} = this.props;

    switch (activeTab) {
      case TabType.OVERVIEW:
        return <MovieOverview
          ratingScore={ratingScore}
          ratingCount={ratingCount}
          description={description}
          director={director}
          starring={starring}
        />;

      case TabType.DETAILS:
        return <MovieDetails
          director={director}
          starring={starring}
          genre={genre}
          year={year}
          runTime={runTime}
        />;

      case TabType.REVIEWS:
        return <MovieReviews />;

      default:
        return null;
    }
  }

  _getSimilarMovies(movies, genre) {
    return movies.filter((movie) => movie.genre === genre).slice(0, MAX_SIMILAR_MOVIES);
  }

  render() {
    const {movie: {id, background, title, genre, year, posterBig, isFavorite}, renderTabs, isLoadingFavoriteMovie, loadMovies} = this.props;
    const similarMovies = this._getSimilarMovies(this.props.movies, genre);

    if (isLoadingFavoriteMovie) {
      loadMovies();
    }

    return (
      <Fragment>
        <section className="movie-card movie-card--full">
          <div className="movie-card__hero">
            <div className="movie-card__bg">
              <img src={background} alt={title}/>
            </div>
            <h1 className="visually-hidden">WTW</h1>
            <Header />
            <div className="movie-card__wrap">
              <div className="movie-card__desc">
                <h2 className="movie-card__title">{title}</h2>
                <p className="movie-card__meta">
                  <span className="movie-card__genre">{genre}</span>
                  <span className="movie-card__year">{year}</span>
                </p>
                <div className="movie-card__buttons">
                  <Link to={`${AppRoute.PLAYER}/${id}`} className="btn btn--play movie-card__button" type="button">
                    <svg viewBox="0 0 19 19" width="19" height="19">
                      <use xlinkHref="#play-s"/>
                    </svg>
                    <span>Play</span>
                  </Link>
                  <AddMyList
                    id={id}
                    isFavorite={isFavorite}
                  />
                  <Link
                    to={`${AppRoute.FILM}/${id}${AppRoute.ADD_REVIEW}`}
                    className="btn movie-card__button">
                    Add review
                  </Link>
                </div>
              </div>
            </div>
          </div>
          <div className="movie-card__wrap movie-card__translate-top">
            <div className="movie-card__info">
              <div className="movie-card__poster movie-card__poster--big">
                <img src={posterBig} alt={title} width="218" height="327"/>
              </div>
              <div className="movie-card__desc">
                {renderTabs()}
                {this._renderActiveTab()}
              </div>
            </div>
          </div>
        </section>
        <div className="page-content">
          <section className="catalog catalog--like-this">
            <h2 className="catalog__title">More like this</h2>
            <MoviesListWrapped
              movies={similarMovies}
            />
          </section>
          <Footer />
        </div>
      </Fragment>
    );
  }
}


MoviePage.propTypes = {
  movie: PropTypes.shape({
    id: PropTypes.number.isRequired,
    background: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    genre: PropTypes.string.isRequired,
    year: PropTypes.number.isRequired,
    poster: PropTypes.string.isRequired,
    posterBig: PropTypes.string.isRequired,
    ratingScore: PropTypes.number.isRequired,
    ratingCount: PropTypes.number.isRequired,
    description: PropTypes.string.isRequired,
    director: PropTypes.string.isRequired,
    starring: PropTypes.arrayOf(
        PropTypes.string.isRequired
    ).isRequired,
    isFavorite: PropTypes.bool.isRequired,
    runTime: PropTypes.number.isRequired,
  }),
  movies: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        title: PropTypes.string.isRequired,
        poster: PropTypes.string.isRequired
      })
  ).isRequired,
  renderTabs: PropTypes.func.isRequired,
  activeTab: PropTypes.string.isRequired,
  setActiveMovie: PropTypes.func.isRequired,
  isLoadingFavoriteMovie: PropTypes.bool.isRequired,
  loadMovies: PropTypes.func.isRequired
};

const mapStateToProps = (state, props) => ({
  movie: getSelectedMovie(state, props.id),
  movies: getMovies(state),
  authorizationStatus: getAuthorizationStatus(state),
  isLoadingFavoriteMovie: getLoadingFavoriteMovie(state)
});

const mapDispatchToProps = (dispatch) => ({
  setActiveMovie(movie) {
    dispatch(ActionCreator.getActiveMovie(movie));
  },

  loadMovies() {
    dispatch(DataOperation.loadMovies());
  }
});


export default connect(mapStateToProps, mapDispatchToProps)(MoviePage);
