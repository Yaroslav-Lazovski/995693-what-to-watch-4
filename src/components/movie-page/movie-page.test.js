import React from "react";
import renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import {Provider} from "react-redux";
import {Router} from "react-router-dom";

import history from "../../history";
import {MoviePage} from "./movie-page.jsx";

import NameSpace from "../../reducer/name-space";
import movies from "../../mocks/films.js";
import {GENRES, AuthorizationStatus} from "../../consts.js";


const movie = {
  id: 1,
  background: `img/movie.jpg`,
  title: `The movie`,
  genre: `Drama`,
  year: 2014,
  poster: `img/movie-poster.jpg`,
  posterBig: `img/movie-poster-big.jpg`,
  ratingScore: 8.9,
  ratingCount: 240,
  description: `Movie description`,
  director: `Movie director`,
  starring: [`Bill Murray`, `Edward Norton`, `Jude Law`, `Willem Dafoe`],
  runTime: 120,
  isFavorite: false
};


const activeTab = `Overview`;

const mockStore = configureStore([]);

const store = mockStore({
  [NameSpace.DATA]: {
    movies
  },
  [NameSpace.STATE]: {
    genre: GENRES.ALL,
    showedMoviesNumber: 8,
  },
  [NameSpace.USER]: {
    authorizationStatus: AuthorizationStatus.NO_AUTH
  }
});


describe(`MoviePage Snapshot`, () => {
  it(`Should movie page render correctly`, () => {
    const tree = renderer
    .create(
        <Router
          history={history}
        >
          <Provider store={store}>
            <MoviePage
              movie={movie}
              movies={movies}
              activeTab={activeTab}
              authorizationStatus={AuthorizationStatus.NO_AUTH}
              renderTabs={() => {}}
              setActiveMovie={() => {}}
              isLoadingFavoriteMovie={false}
              loadMovies={() => {}}
            />
          </Provider>
        </Router>, {
          createNodeMock: ()=>{
            return {};
          }
        })
    .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
