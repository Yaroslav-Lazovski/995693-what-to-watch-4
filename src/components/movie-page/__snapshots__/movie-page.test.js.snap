// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoviePage Snapshot Should movie page render correctly 1`] = `
Array [
  <section
    className="movie-card movie-card--full"
  >
    <div
      className="movie-card__hero"
    >
      <div
        className="movie-card__bg"
      >
        <img
          alt="The movie"
          src="img/movie.jpg"
        />
      </div>
      <h1
        className="visually-hidden"
      >
        WTW
      </h1>
      <header
        className="page-header movie-card__head"
      >
        <div
          className="logo"
        >
          <a
            className="logo__link"
            href="/"
            onClick={[Function]}
          >
            <span
              className="logo__letter logo__letter--1"
            >
              W
            </span>
            <span
              className="logo__letter logo__letter--2"
            >
              T
            </span>
            <span
              className="logo__letter logo__letter--3"
            >
              W
            </span>
          </a>
        </div>
        <div
          className="user-block"
        >
          <a
            className="user-block__link"
            href="/login"
            onClick={[Function]}
          >
            Sign in
          </a>
        </div>
      </header>
      <div
        className="movie-card__wrap"
      >
        <div
          className="movie-card__desc"
        >
          <h2
            className="movie-card__title"
          >
            The movie
          </h2>
          <p
            className="movie-card__meta"
          >
            <span
              className="movie-card__genre"
            >
              Drama
            </span>
            <span
              className="movie-card__year"
            >
              2014
            </span>
          </p>
          <div
            className="movie-card__buttons"
          >
            <a
              className="btn btn--play movie-card__button"
              href="/player/1"
              onClick={[Function]}
              type="button"
            >
              <svg
                height="19"
                viewBox="0 0 19 19"
                width="19"
              >
                <use
                  xlinkHref="#play-s"
                />
              </svg>
              <span>
                Play
              </span>
            </a>
            <button
              className="btn btn--list movie-card__button"
              onClick={[Function]}
              type="button"
            >
              <svg
                height="20"
                viewBox="0 0 19 20"
                width="19"
              >
                <use
                  xlinkHref="#add"
                />
              </svg>
              <span>
                My list
              </span>
            </button>
            <a
              className="btn movie-card__button"
              href="/films/1/review"
              onClick={[Function]}
            >
              Add review
            </a>
          </div>
        </div>
      </div>
    </div>
    <div
      className="movie-card__wrap movie-card__translate-top"
    >
      <div
        className="movie-card__info"
      >
        <div
          className="movie-card__poster movie-card__poster--big"
        >
          <img
            alt="The movie"
            height="327"
            src="img/movie-poster-big.jpg"
            width="218"
          />
        </div>
        <div
          className="movie-card__desc"
        >
          <div
            className="movie-rating"
          >
            <div
              className="movie-rating__score"
            >
              8,9
            </div>
            <p
              className="movie-rating__meta"
            >
              <span
                className="movie-rating__level"
              >
                Very good
              </span>
              <span
                className="movie-rating__count"
              >
                240
                 ratings
              </span>
            </p>
          </div>
          <div
            className="movie-card__text"
          >
            <p>
              Movie description
            </p>
            <p
              className="movie-card__director"
            >
              <strong>
                Director: 
                Movie director
              </strong>
            </p>
            <p
              className="movie-card__starring"
            >
              <strong>
                Starring: 
                Bill Murray, Edward Norton, Jude Law, Willem Dafoe
                 and other
              </strong>
            </p>
          </div>
        </div>
      </div>
    </div>
  </section>,
  <div
    className="page-content"
  >
    <section
      className="catalog catalog--like-this"
    >
      <h2
        className="catalog__title"
      >
        More like this
      </h2>
      <div
        className="catalog__movies-list"
      >
        <article
          className="small-movie-card catalog__movies-card"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <a
            href="/films/3"
            onClick={[Function]}
          >
            <div
              className="small-movie-card__image"
            >
              <video
                className="player__video"
                poster="img/macbeth.jpg"
                src="https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4"
              />
              <img
                alt="Macbeth"
                height="175"
                src="img/macbeth.jpg"
                width="280"
              />
            </div>
          </a>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="/films/3"
              onClick={[Function]}
            >
              Macbeth
            </a>
          </h3>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <a
            href="/films/4"
            onClick={[Function]}
          >
            <div
              className="small-movie-card__image"
            >
              <video
                className="player__video"
                poster="img/aviator.jpg"
                src="https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm"
              />
              <img
                alt="Aviator"
                height="175"
                src="img/aviator.jpg"
                width="280"
              />
            </div>
          </a>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="/films/4"
              onClick={[Function]}
            >
              Aviator
            </a>
          </h3>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <a
            href="/films/5"
            onClick={[Function]}
          >
            <div
              className="small-movie-card__image"
            >
              <video
                className="player__video"
                poster="img/we-need-to-talk-about-kevin.jpg"
                src="https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4"
              />
              <img
                alt="We need to talk about Kevin"
                height="175"
                src="img/we-need-to-talk-about-kevin.jpg"
                width="280"
              />
            </div>
          </a>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="/films/5"
              onClick={[Function]}
            >
              We need to talk about Kevin
            </a>
          </h3>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <a
            href="/films/7"
            onClick={[Function]}
          >
            <div
              className="small-movie-card__image"
            >
              <video
                className="player__video"
                poster="img/revenant.jpg"
                src="https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4"
              />
              <img
                alt="Revenant"
                height="175"
                src="img/revenant.jpg"
                width="280"
              />
            </div>
          </a>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="/films/7"
              onClick={[Function]}
            >
              Revenant
            </a>
          </h3>
        </article>
      </div>
    </section>
    <footer
      className="page-footer"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link logo__link--light"
          href="/"
          onClick={[Function]}
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="copyright"
      >
        <p>
          Â© 2019 What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>,
]
`;
